File: solutions/clojure/largest-series-product/1/src/largest_series_product.clj
(ns largest-series-product)

(defn largest-product [span s]
  (cond
    (neg? span) (throw (IllegalArgumentException. "span must not be negative"))
    (> span (count s)) (throw (IllegalArgumentException. "span must not exceed string length"))
    (re-find #"\D" s) (throw (IllegalArgumentException. "digits input must only contain digits"))
    :else (->> s
               (map #(Character/digit % 10))
               (partition span 1)
               (map #(reduce * %))
               (apply max))))

File: solutions/ruby/largest-series-product/1/largest_series_product.rb
# frozen_string_literal: true

# class Series
class Series
  def initialize(digits) = digits.match?(/^\d*$/)
    && (@digits = digits.chars.map(&:to_i))
    || raise(ArgumentError)

  def largest_product(span) = (0..@digits.length).cover?(span)
    && @digits.each_cons(span).map { _1.reduce(:*) }.max
    || raise(ArgumentError)
end
